### PR TITLE
refactor!: Merge `CustomOp` and `ExternalOp`.

### PR DESCRIPTION
**Pull request by aborgna-q**
Followup of #922 .

Drops `ExternalOp` and makes `CustomOp` an enum with two boxed variants for `OpaqueOp` and `ExtensionOp`.
The schema remains the same.

### CODE REVIEW

File: hugr/src/hugr/serialize.rs [hunk 1]
```diff
@@ -349,10 +349,7 @@ pub mod test {
             let old_op = h_canon.get_optype(node);
             if let OpType::CustomOp(new_op) = new_op {
                 if let OpType::CustomOp(old_op) = old_op {
-                    assert_eq!(
-                        new_op.as_ref().clone().into_opaque(),
-                        old_op.as_ref().clone().into_opaque()
-                    );
+                    assert_eq!(new_op.clone().into_opaque(), old_op.clone().into_opaque());
```
Comment (ss2165) [inline hugr/src/hugr/serialize.rs:352]: not doing the "into_opaque" here could address the lack of coverage of PartialEq in the other PR. Maybe that change should be in this PR? 

I do worry if that PartialEq implementation is a good idea though, seems likely to induce hard to find bugs.

**Review by ss2165 — APPROVED** [review]
nice 👍

File: hugr/src/hugr/serialize.rs (hunk 1 above)
Comment (aborgna-q) [reply hugr/src/hugr/serialize.rs]: The main idea of the `PartialEq` impl is that we can replace this hacky test with just `==`
https://github.com/CQCL/hugr/pull/923/commits/993629457b564cca1eec475f7b42ae779ad71463

In general, a serialization roundtrip should preserve equality. Resolving / unresolving a custom operation is an internal operation that doesn't impact the semantics of the hugr.

Comment (ss2165) [reply hugr/src/hugr/serialize.rs]: ok I'm convinced, but scared

Comment (aborgna-q) [reply hugr/src/hugr/serialize.rs]: Perhaps I should add some `is_opaque`/`is_extension` methods?

**Pull request merged**